File: materials/CompareAlgorithms/vertical_drift/algorithms/Matlab/cluster.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% CLUSTER
%
% https://github.com/sascha2schroeder/popEye/
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

function fixation_XY = cluster(fixation_XY, line_Y)

	n = size(fixation_XY, 1);
	m = length(line_Y);
	fixation_Y = fixation_XY(:, 2);
	[clusters, centers] = kmeans(fixation_Y, m);
	[~, ordered_cluster_indices] = sort(centers);
	for fixation_i = 1 : n
		cluster_i = clusters(fixation_i);
		line_i = find(ordered_cluster_indices == cluster_i);
		fixation_XY(fixation_i, 2) = line_Y(line_i);
	end

end

File: R/Cluster.R

# reimplementation from Carr et al. used

Cluster <- function(fixation_XY, line_Y) {
  n <- nrow(fixation_XY)
  m <- length(line_Y)
  fixation_Y <- fixation_XY[, 2]
  clusters <- kmeans(fixation_Y, m, iter.max=300, nstart=100)
  ordered_cluster_indices <- order(clusters$centers)
  for (fixation_i in 1 : n) {
    cluster_i <- clusters$cluster[fixation_i]
    line_i <- which(ordered_cluster_indices == cluster_i)
    fixation_XY[fixation_i, 2] <- line_Y[line_i]
  }
  return(fixation_XY)
}
